### PR TITLE
APEXMALHAR-2284 *Review Only* 1) Implemented SpillableMap, SpillableArrayList, SpillableArrayListMultiMap over TimeSlicedBucketState 2) Integrated Spillable data structure into Inner Join operator

### PR DESCRIPTION
**Pull request by chaithu14**
1) Implemented SpillableMap, SpillableArrayList, SpillableArrayListMultiMap over TimeSlicedBucketState 2) Integrated Spillable data structure into Inner Join operator

### CODE REVIEW

**Comment by chaithu14** [comment]
@chandnisingh @bhupeshchawda @tweise Please review the PR.

File: library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java [hunk 1]
```diff
@@ -245,4 +274,46 @@ public FieldObjectMap()
...
+      if (timeFieldGet == null) {
+        Class timeField = null;
+        try {
+          timeField = ClassUtils.primitiveToWrapper(o.getClass().getDeclaredField(timeFieldExpr).getType());
+        } catch (NoSuchFieldException e) {
+          throw new RuntimeException(e);
+        }
+        timeFieldGet = PojoUtils.createGetter(o.getClass(), timeFieldExpr, timeField);
+      }
+      return (long)timeFieldGet.get(o);
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java]: Typecasting an object to long might not be a good idea.

File: library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java [hunk 2]
```diff
@@ -245,4 +274,46 @@ public FieldObjectMap()
...
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public long getTime(Object o)
+    {
+      if (timeFieldGet == null) {
+        Class timeField = null;
+        try {
+          timeField = ClassUtils.primitiveToWrapper(o.getClass().getDeclaredField(timeFieldExpr).getType());
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java]: I'll suggest to have a default implementation for for the time field as `Date` and create a getter for that.
Do not have a generic getter. We don't know what type the configured date field has.

File: library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java [hunk 3]
```diff
@@ -245,4 +274,46 @@ public FieldObjectMap()
...
+    public long getBucket(Object o)
+    {
+      return o.hashCode() % getNoOfBuckets();
+    }
+  }
+
+  /**
+   * Extract the time value from Object
+   */
+  public class POJOExtractTime implements TimeExtractor<Object>
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/join/POJOInnerJoinOperator.java]: Rename to `POJOTimeExtractor`?

File: library/src/main/java/org/apache/apex/malhar/lib/join/AbstractManagedStateInnerJoinOperator.java [hunk 4]
```diff
@@ -90,17 +113,36 @@ public void createStores()
...
+  @SuppressWarnings("unchecked")
   @Override
   protected void processTuple(T tuple, boolean isStream1Data)
   {
     Spillable.SpillableListMultimap<K,T> store = isStream1Data ? stream1Data : stream2Data;
     K key = extractKey(tuple,isStream1Data);
-    long timeBucket = extractTime(tuple,isStream1Data);
-    if (!((ManagedTimeStateMultiValue)store).put(key, tuple,timeBucket)) {
+    //long timeBucket = extractTime(tuple,isStream1Data);
+    if (!store.put(key, tuple)) {
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/join/AbstractManagedStateInnerJoinOperator.java]: Please mention why `put` would return false? Seems like all implementations return true by default.

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableTimeStateStore.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.apex.malhar.lib.state.TimeSlicedBucketedState;
+import org.apache.apex.malhar.lib.state.managed.BucketProvider;
+import org.apache.hadoop.classification.InterfaceStability;
+
+import com.datatorrent.api.Component;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Operator;
+
+@InterfaceStability.Evolving
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableTimeStateStore.java:32]: Please add java docs explaining use of this interface.

**Review by bhupeshchawda — COMMENTED** [review]
I have some very initial comments. I haven't yet looked at all the changes.
Will add more comments soon.

**Comment by bhupeshchawda** [comment]
@chaithu14 Can you update the description of the PR?

File: library/src/main/java/org/apache/apex/malhar/lib/join/AbstractManagedStateInnerJoinOperator.java (hunk 4 above)
Comment (chaithu14) [reply library/src/main/java/org/apache/apex/malhar/lib/join/AbstractManagedStateInnerJoinOperator.java]: Yes you are right. Before inserting, need to check whether the tuple is expired or not. For this check, I added isExpiredTuple() API.

**Comment by chaithu14** [comment]
Incorporated review comments. Added Unit test cases for new spillable structures and updated PR.

**Comment by tweise** [comment]
Why do we need new spillable DS for this? Very similar work was done for WindowOperator recently, why can't it be reused? It may also be useful to call out the difference for the join functionality itself compared to what WindowOperator already supports. @davidyan74 any thoughts?

**Comment by chaithu14** [comment]
@tweise Innerjoin operator uses the ManagedTimeStateImpl store which has the key bucket and time bucket as we need to put(keyBucket, timeBucket, key, value) call.
Existing spillable structures uses the BucketedState which has the below calls
             - put(bucketId, key, value) 
             - getSync(bucketId, key)  
             - getAsync(bucketId, key) where bucketId is either key bucket or time bucket. 
So, existing spillable structures can't be reused for InnerJoin operator.

**Comment by davidyan74** [comment]
@chaithu14: Work is in progress to use ManagedTimeStateImpl for the spillable data structures. The work is currently being done by @siyuanh and it's recorded in this umbrella ticket https://issues.apache.org/jira/browse/APEXMALHAR-2130

**Comment by davidyan74** [comment]
@tweise Yes, this PR looks like most of it we are already working on so I feel a bit uneasy about this PR because of duplication of similar work. 

About your comment about the join functionality between this and WindowedOperator, the current innerjoin operator does not require the other concepts required by the WindowedOperator (watermark, triggering etc) so that may be a reason to continue development on the inner join operator. Whether it's a big enough reason or not is debatable. 

My take is that we should halt further development of the join operator and evaluate the WindowedMergeOperator and see if there is anything missing from the current use cases of the join operator.

**Comment by tweise** [comment]
@davidyan74 this is blocking the release. But APEXMALHAR-2130 was also supposed to be finished for 3.6.0. What is that status, can it be completed and used to make this join operator functional until we decide about longer term plan?

**Comment by davidyan74** [comment]
OK, I'm asking @siyuanh to update the status of the affected tickets.

**Comment by chaithu14** [comment]
@davidyan74 Are you talking about this PR (#345)? If yes, then the duplication is very minimal and I have seen only the interface (SpillableTimeStateStore) is common. 
Spillable structures in this PR are over the TimeSlicedBucketedState and timeExtractor be the time extract from value.

**Comment by ilganeli** [comment]
@chaithu14 Is this work still relevant or can this be closed?

**Pull request closed without merging**